runner.go

```diff
@@ -116,8 +182,14 @@ var ErrDead = errors.New("worker runner is not running")
 //
 // StartWorker returns ErrDead if the runner is not running.
 func (runner *Runner) StartWorker(id string, startFunc func() (Worker, error)) error {
+	// Note: we need the reply channel so that when StartWorker
+	// returns, we're guaranteed that the worker is installed
+	// when we return, so Worker will see it if called
+	// immediately afterwards.
+	reply := make(chan struct{})
 	select {
-	case runner.startc <- startReq{id, startFunc}:
+	case runner.startc <- startReq{id, startFunc, reply}:
+		<-reply
```

Comment:
It's OK because the startc channel is synchronous so if we succeed in sending on it, we know that the run goroutine has entered the startc arm of the select, and that calls startWorker which never blocks (trivially verifiable) and then closes the reply channel, so we can be completely sure we'll always get a reply.